packages/common/core-utils/src/assert.ts

```diff
@@ -22,3 +22,87 @@ export function assert(condition: boolean, message: string | number): asserts co
 		);
 	}
 }
+
+/**
+ * Asserts that can be conditionally enabled in debug/development builds but will be optimized out of production builds.
+ *
+ * @param predicate - A pure function that should return true if the condition holds, or a string or object describing the condition that failed.
+ * This function will only be run in some configurations so it should be pure, and only used to detect bugs (when debugAssert are enabled), and must not be relied on to enforce the condition is true: for that use {@link assert}.
+ * @remarks
+ * Remarks optimizing the asserts out of the bundle requires a bundler like webpack which leverages `__PURE__` annotations like https://webpack.js.org/guides/tree-shaking/#mark-a-function-call-as-side-effect-free.
+ *
+ * Exceptions thrown by this function must never be caught in production code as that will result in different behavior when testing and when running optimized builds.
+ * The `predicate` function must be pure (have no side-effects) to ensure that the behavior of code is the same regardless of if the asserts are disabled, enabled or optimized out.
+ *
+ * These asserts are disabled by default, even in debug builds to ensure that by default code will be tested as production runs, with them disabled.
+ * Additionally this ensures that apps that use a bundler which does not remove `__PURE__` will not incur the runtime cost of calling the predicate.
+ * These asserts can be can be enabled by calling `configureDebugAsserts(true)`: see {@link configureDebugAsserts}.
+ *
+ * @privateRemarks
+ * This design was chosen to accomplish two main goals:
+ *
+ * 1. Make it easy to compile debug asserts fully out of production builds.
+ * For webpack this happens by default, avoiding the need for customers to do special configuration.
+ * This is important for both performance and bundle size.
+ * 2. Make it easy to test (both manually and automated) with and without the predicates running.
+ * This ensures it is possible to benefit from the asserts when enabled, but also test with them disabled to ensure this disablement doesn't cause bugs.
+ *
+ * The default behavior of having debugAsserts disabled helps ensure that tests which don't know about debug asserts will still run in a way that is most similar to production.
```

Comment:
I'm not sure how I feel about the choose to disable them by default. I think it might be the right one, but maybe we want to make our fluid test utils enable them by default, then where relevant tests can opt out?

Maybe we can have a CI job that runs the test suite both ways (and tag specific tests expected to fail with that which we filter out?). THis is a subjective choice, and there are good reasons to go either way.

Reply:
Since these are off by default, which is not normally how "debug" assert libraries work, maybe we should call it "expensiveAssert", or maybe we should have both, and make this on by default? Or maybe we want both and enable this one in our test utils by default?

As this is internal, we can evolve its usage over time, but maybe having a couple of different options will help developers encode more semantics into the use when they add them.

Reply:
Maybe calling this "developmentOnlyAssert" and making it on by default would be good. Then we could recommend the pattern of writing them like `developmentOnlyAssert(() => expensiveAsserts && expensiveCheck)` or developmentOnlyAssert(() => testAsserts && check)` if you want them disabled by default (use your own flags).

Reply:
Personally, I think keeping the recommended usage pattern simple would be best, so I would vote against the latter pattern myself.

I do think a different name would probably be better. I can see good arguments for both `expensiveAssert` and `developmentOnlyAssert`. The latter is probably the most accurate, though the former perhaps does a better job at hinting at the expected use case. Neither makes explicit that they are disabled by default (assuming that remains the default), but the latter maybe hints at that better?